File: YeetMacro2/Resources/Raw/MacroSets/eversoul/scripts/doOutings.js
// @position=11
// Do all outings. Required setting: targetSoul (pattern that selects soul)
// If targetBondLevel is checked and the value is reached, script will stop
const resolution = macroService.GetCurrentResolution();
const loopPatterns = [patterns.lobby.level, patterns.town.level, patterns.town.outings.outingsCompleted, patterns.town.outings, patterns.titles.outingGo, patterns.town.outings.selectAKeyword];
const daily = dailyManager.GetDaily();

if (daily.doOutings.done.IsChecked) {
	return "Script already completed. Uncheck done to override daily flag.";
}
const targetBondLevelIsEnabled = settings.doOutings.targetBondLevel.IsEnabled;
const targetBondLevel = settings.doOutings.targetBondLevel.Value;
const targetSoul = macroService.ClonePattern(settings.doOutings.targetSoul.Value, {
	X: 275,
	Y: 80,
	Width: resolution.Width - 915,
	Height: 900,
	Path: 'settings.outings.target',
	OffsetCalcType: 'DockLeft'
});

const maxSwipes = 5;
let swipeCount = 0;

while (macroService.IsRunning) {
	const loopResult = macroService.PollPattern(loopPatterns);
	switch (loopResult.Path) {
		case 'lobby.level':
			logger.info('doOutings: click town');
			macroService.PollPattern(patterns.lobby.town, { DoClick: true, PredicatePattern: patterns.town.enter });
			sleep(500);
			macroService.PollPattern(patterns.town.enter, { DoClick: true, ClickPattern: patterns.general.tapTheScreen, PredicatePattern: patterns.town.level });
			break;
		case 'town.level':
			logger.info('doOutings: click info with Offset');
			macroService.PollPattern(patterns.town.info, { DoClick: true, ClickOffset: { X: -60 }, PredicatePattern: patterns.town.outings });
			break;
		case 'town.outings':
			logger.info('doOutings: click outing target');
			sleep(500);
			swipeCount = 0;
			let result = macroService.PollPattern(targetSoul, { TimeoutMs: 2_000 });
			while (macroService.IsRunning && !result.IsSuccess && swipeCount < maxSwipes) {
				macroService.DoSwipe({ X: 1080, Y: 800 }, { X: 1080, Y: 250 });
				sleep(500);
				result = macroService.PollPattern(targetSoul, { TimeoutMs: 2_000 });
				swipeCount++;
			}
			if (!result.IsSuccess) {
				return 'Unable to find target soul';
			} 
			macroService.ClickPattern(targetSoul);
			macroService.PollPattern(targetSoul, { DoClick: true, PredicatePattern: patterns.town.outings.call });

			if (targetBondLevelIsEnabled) {
				sleep(1_000);
				const currentBondLevel = Number.parseInt(macroService.GetText(patterns.town.outings.bondLevel)?.replace(/[ ]/g, ''));
				if (currentBondLevel >= Number(targetBondLevel)) {
					throw new Error(`Target bond level reached: ${targetBondLevel}`);
				}
			}
			
			sleep(500);
			macroService.PollPattern(patterns.town.outings.call, { DoClick: true, PredicatePattern: patterns.prompt.confirm });
			sleep(500);
			macroService.PollPattern(patterns.prompt.confirm, { DoClick: true, PredicatePattern: patterns.town.outings.goOnOuting });
			sleep(500);
			macroService.PollPattern(patterns.town.outings.goOnOuting, { DoClick: true, ClickPattern: patterns.prompt.next, PredicatePattern: patterns.titles.outingGo });
			break;
		case 'titles.outingGo':
			logger.info('doOutings: select outing');
			const outingNumber = 1 + Math.floor(Math.random() * 5);
			logger.debug('outingNumber: ' + outingNumber);
			macroService.PollPattern(patterns.town.outings['outing' + outingNumber], { DoClick: true, PredicatePattern: patterns.prompt.confirm });
			sleep(500);
			macroService.PollPattern(patterns.prompt.confirm, { DoClick: true, ClickPattern: patterns.prompt.next, PredicatePattern: patterns.town.outings.keywordSelectionOpportunity });
			sleep(500);
			macroService.PollPattern(patterns.prompt.next, { DoClick: true, ClickPattern: patterns.prompt.middleTap, PredicatePattern: patterns.town.outings.selectAKeyword });
			break;
		case 'town.outings.selectAKeyword':
			logger.info('doOutings: select keyword');
			const keywordPoints1 = macroService.GetText(patterns.town.outings.keywordPoints1).replace(/[\+ ]/g, '');
			const keywordPoints2 = macroService.GetText(patterns.town.outings.keywordPoints2).replace(/[\+ ]/g, '');
			const keywordPoints3 = macroService.GetText(patterns.town.outings.keywordPoints3).replace(/[\+ ]/g, '');
			logger.info('keywordPoints1: ' + keywordPoints1);
			logger.info('keywordPoints2: ' + keywordPoints2);
			logger.info('keywordPoints3: ' + keywordPoints3);
			const keywordPoints = [Number(keywordPoints1), Number(keywordPoints2), Number(keywordPoints3)];
			const maxIdx = keywordPoints.reduce((maxIdx, val, idx, arr) => val > arr[maxIdx] ? idx : maxIdx, 0);
			logger.debug('keywordTarget: ' + (maxIdx + 1));
			macroService.PollPattern(patterns.town.outings['keywordPoints' + (maxIdx + 1)], { DoClick: true, PredicatePattern: patterns.prompt.next });
			sleep(500);
			const promptNextResult = macroService.PollPattern(patterns.prompt.next, { DoClick: true, ClickPattern: [patterns.prompt.next, patterns.prompt.tapTheScreen, patterns.prompt.middleTap], PredicatePattern: [patterns.town.outings.selectAKeyword, patterns.town.outings] });
			if (macroService.IsRunning && promptNextResult.PredicatePath === 'town.outings') {
				daily.doOutings.count.Count++;
			}
			break;
		case 'town.outings.outingsCompleted':
			if (macroService.IsRunning) {
				daily.doOutings.done.IsChecked = true;
			}
			return;
	}

	sleep(1_000);
}

File: YeetMacro2/Services/Scripts/Eversoul/doOutings.cs
using Microsoft.Extensions.Logging;
using System.Text.RegularExpressions;
using YeetMacro2.Data.Models;
using YeetMacro2.ViewModels.NodeViewModels;

namespace YeetMacro2.Services.Scripts.Eversoul;

public partial class EversoulScripts
{
    public string doOutings()
    {
        var done = false;
        var loopPatterns = new PatternNode[] { patterns["lobby"]["everstone"], patterns["town"]["evertalk"], patterns["town"]["outings"]["outingsCompleted"], patterns["town"]["outings"], patterns["titles"]["outingGo"], patterns["town"]["outings"]["selectAKeyword"] };
        var targetSoul = PatternNodeManagerViewModel.CloneNode(settings["outings"]["target"].GetValue<PatternNode>());
        targetSoul.Path = "settings.outings.target";
        foreach (var p in targetSoul.Patterns)
        {
            // original width: 1005.4752807617188 (should calculate using resolution comparisons instead)
            p.Rect = new Rect(275.85736083984375, 82.9250717163086, 1500.4752807617188, 857.20263671875);
        }
        while (macroService.IsRunning && !done)
        {
            var result = macroService.PollPattern(loopPatterns);
            switch (result.Path)
            {
                case "lobby.everstone":
                    logger.LogInformation("doOutings: click town");
                    macroService.PollPattern(patterns["lobby"]["town"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["town"]["enter"] });
                    new System.Threading.ManualResetEvent(false).WaitOne(500);
                    macroService.PollPattern(patterns["town"]["enter"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["town"]["evertalk"] });
                    break;
                case "town.evertalk":
                    logger.LogInformation("doOutings: click info with Offset");
                    macroService.PollPattern(patterns["town"]["info"], new PollPatternFindOptions() { DoClick = true, Offset = new Point(-60, 0), PredicatePattern = patterns["town"]["outings"] });
                    break;
                case "town.outings":
                    logger.LogInformation("doOutings: click outing target");
                    new System.Threading.ManualResetEvent(false).WaitOne(500);
                    macroService.ClickPattern(targetSoul);
                    macroService.PollPattern(targetSoul, new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["town"]["outings"]["call"] });
                    new System.Threading.ManualResetEvent(false).WaitOne(500);
                    macroService.PollPattern(patterns["town"]["outings"]["call"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["prompt"]["confirm"] });
                    new System.Threading.ManualResetEvent(false).WaitOne(500);
                    macroService.PollPattern(patterns["prompt"]["confirm"], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["town"]["outings"]["goOnOuting"] });
                    new System.Threading.ManualResetEvent(false).WaitOne(500);
                    macroService.PollPattern(patterns["town"]["outings"]["goOnOuting"], new PollPatternFindOptions() { DoClick = true, ClickPattern = patterns["prompt"]["next"], PredicatePattern = patterns["titles"]["outingGo"] });
                    break;
                case "titles.outingGo":
                    logger.LogInformation("doOutings: select outing");
                    var outingNumber = macroService.Random(1, 4);
                    logger.LogDebug("outingNumber: " + outingNumber);

                    macroService.PollPattern(patterns["town"]["outings"]["outing" + outingNumber], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["prompt"]["confirm"] });
                    new System.Threading.ManualResetEvent(false).WaitOne(500);
                    macroService.PollPattern(patterns["prompt"]["confirm"], new PollPatternFindOptions() { DoClick = true, ClickPattern = patterns["prompt"]["next"], PredicatePattern = patterns["town"]["outings"]["keywordSelectionOpportunity"] });
                    new System.Threading.ManualResetEvent(false).WaitOne(500);
                    macroService.PollPattern(patterns["prompt"]["next"], new PollPatternFindOptions() { DoClick = true, ClickPattern = patterns["prompt"]["middleTap"], PredicatePattern = patterns["town"]["outings"]["selectAKeyword"] });
                    break;
                case "town.outings.selectAKeyword":
                    logger.LogInformation("doOutings: select keyword");
                    var keywordPoints1 = Regex.Replace(macroService.GetText(patterns["town"]["outings"]["keywordPoints1"]), "[, ]", "");
                    var keywordPoints2 = Regex.Replace(macroService.GetText(patterns["town"]["outings"]["keywordPoints2"]), "[, ]", "");
                    var keywordPoints3 = Regex.Replace(macroService.GetText(patterns["town"]["outings"]["keywordPoints3"]), "[, ]", "");
                    logger.LogInformation("keywordPoints1: " + keywordPoints1);
                    logger.LogInformation("keywordPoints2: " + keywordPoints2);
                    logger.LogInformation("keywordPoints3: " + keywordPoints3);

                    var keywordPoints = new int[] { int.Parse(keywordPoints1), int.Parse(keywordPoints2), int.Parse(keywordPoints3) };
                    int maxIdx = keywordPoints
                        .Select((val, idx) => new { Value = val, Index = idx })
                        .Aggregate((max, current) => current.Value > max.Value ? current : max)
                        .Index;
                    logger.LogDebug("keywordTarget: " + (maxIdx + 1));
                    macroService.PollPattern(patterns["town"]["outings"]["keywordPoints" + (maxIdx + 1)], new PollPatternFindOptions() { DoClick = true, PredicatePattern = patterns["prompt"]["next"] });
                    new System.Threading.ManualResetEvent(false).WaitOne(500);
                    macroService.PollPattern(patterns["prompt"]["next"], new PollPatternFindOptions() { DoClick = true, ClickPattern = new PatternNode[] { patterns["prompt"]["next"], patterns["prompt"]["tapTheScreen"], patterns["prompt"]["middleTap"] }, PredicatePattern = new PatternNode[] { patterns["town"]["outings"]["selectAKeyword"], patterns["town"]["outings"] } });
                    break;
                case "town.outings.outingsCompleted":
			    done = true;
                break;
            }

            new System.Threading.ManualResetEvent(false).WaitOne(1_000);
        }
        logger.LogInformation("Done...");
        return String.Empty;
    }
}